sn/src/dbs/kv_store/mod.rs

```
@@ -97,96 +95,31 @@ impl<K: Key, V: Value + Send + Sync> KvStore<K, V> {
-    ///
-    /// If a value with the same id already exists, it will be overwritten.
-    #[allow(unused)] // this will soon be used
-    pub(crate) async fn store_batch(&self, values: &[V]) -> Result<()> {
-        info!("Writing batch");
-        use rayon::prelude::*;
-        type KvPair = (String, Vec<u8>);
-        type KvPairResults = Vec<Result<KvPair>>;
-
-        let mut batch = sled::Batch::default();
-        let (ok, err_results): (KvPairResults, KvPairResults) = values
-            .par_iter()
-            .map(|value| {
-                let serialised_value = serialise(value)?.to_vec();
-                let key = value.key().to_db_key()?;
-                Ok((key, serialised_value))
-            })
-            .partition(|r| r.is_ok());
-
-        if !err_results.is_empty() {
-            for e in err_results {
-                error!("{:?}", e);
-            }
 
-            return Err(Error::SledBatching);
-        }
-
-        let consumed_space = ok
-            .into_iter()
-            .flatten()
-            .map(|(key, value)| {
-                let consumed_space = value.len() as u64;
-                batch.insert(key.as_bytes(), value);
-                consumed_space
-            })
-            .sum();
-
-        if !self.used_space.can_consume(consumed_space).await {
+        let exists = self.db.contains_key(key.clone()).map_err(Error::from)?;
+        if !exists
+            && !self
+                .used_space
+                .can_consume(serialised_value.len() as u64)
+                .await
+        {
             return Err(Error::NotEnoughSpace);
         }
 
-        let res = self.db.apply_batch(batch);
+        // Atomically write the value if it's new - this prevents multiple concurrent writes from
+        // consuming extra space (since sled is backed by a log). We throw away the value otherwise,
+        // since current use-cases do not require updates.
+        let _res = self
```

We could maybe log the result as before in the `match` below? Just to be wary that we aren't ignoring an error.

---

Yeah, I swithered on what logging to include. I think it's a good point to log it/match on it though, to make sure we don't lose the `?` in a refactor.